### ISSUE TITLE
SA1015: infinite loop detection not working for closures

### ISSUE THREAD

**Issue by FiloSottile** [dominikh/go-staticcheck#62]
An endless `for` ranging over a `time.Tick` is likely fine (unless the loop blocks for some reason, but there are more leaks in that case).

```go
for range time.Tick(1 * time.Second) {
    s.ReqIncomingSize.Update(int64(incoming.Length()))
}
```

**Comment by dominikh** [comment]
You'll have to expand on your initial comment. Are you saying that this is code we should flag but aren't flagging? Or that we're flagging it but shouldn't be? 

We only report the leaky nature of `time.Tick` in functions that are theoretically able to return and aren't in package main or tests. Ranging over the channel returned by `time.Tick` is known to never return on its own, so unless there is a possible code path that skips the loop entirely, or breaks out of the loop, the use of `time.Tick` shouldn't get flagged in that function.

**Comment by FiloSottile** [comment]
This is code that got flagged but AFAICT shouldn't. There is nothing that can break out of the loop without a panic. This is the entire function verbatim:

```go
    go func() {
        for _ = range time.Tick(1 * time.Second) {
            s.ReqIncomingSize.Update(int64(incoming.Length()))
        }
    }()
```

**Comment by dominikh** [comment]
The issue is that our loop detection isn't working correctly for closures. Will fix.

**Comment by dominikh** [comment]
Fixed.
